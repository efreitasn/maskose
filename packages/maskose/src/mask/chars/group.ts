import { MaskoseCharBase, MaskoseChar } from '.';

export const MASKOSE_CHAR_GROUP_TYPE = 'MASKOSE_CHAR_GROUP_TYPE';
export const MASKOSE_CHAR_GROUP_STR_ID = 'GROUP';

export interface MaskoseCharGroup extends MaskoseCharBase {
  type: typeof MASKOSE_CHAR_GROUP_TYPE;
  chars: MaskoseChar[];
};

/**
 * A group of characters
 */
export default function mkCharGroup(chars: MaskoseChar[]): MaskoseCharGroup {
  if (chars.length === 0) {
    throw new Error('The provided chars array must not be empty.');
  }

  return {
    type: MASKOSE_CHAR_GROUP_TYPE,
    repetitions: 1,
    chars,
    maskedValueLengthConditions: [],
    valueToBeMaskedLengthConditions: []
  };
}