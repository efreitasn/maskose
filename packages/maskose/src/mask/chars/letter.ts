import { MaskoseMaskCharBase } from '.';

export const MASKOSE_CHAR_LETTER_TYPE = 'MASKOSE_CHAR_LETTER_TYPE';
export const MASKOSE_CHAR_LETTER_STR_ID = 'LETTER';

export interface MaskoseMaskCharLetter extends MaskoseMaskCharBase {
  type: typeof MASKOSE_CHAR_LETTER_TYPE;
  regExp: RegExp;
};

interface MkCharLetterOptions {
  /**
   * Whether the range will be case sensitive, defaults to false
   */
  caseSensitive?: boolean;
}

/**
 * A character in the range A-Z expected to be in the value to be masked
 */
export default function mkCharLetter(options: MkCharLetterOptions = {}): MaskoseMaskCharLetter {
  const regExp = new RegExp(
    '[A-Z]',
    options.caseSensitive ? '' : 'i'
  );
  
  return {
    type: MASKOSE_CHAR_LETTER_TYPE,
    repetitions: 1,
    maskedValueLengthConditions: [],
    valueToBeMaskedLengthConditions: [],
    regExp
  };
}