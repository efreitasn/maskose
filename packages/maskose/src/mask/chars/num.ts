import { MaskoseCharBase } from '.';

export const MASKOSE_CHAR_NUM_TYPE = 'MASKOSE_CHAR_NUM_TYPE';
export const MASKOSE_CHAR_NUM_STR_ID = 'NUM';

export interface MaskoseCharNum extends MaskoseCharBase {
  type: typeof MASKOSE_CHAR_NUM_TYPE;
  regExp: RegExp;
};

/**
 * A character in the range 0 to 9 expected to be in the value to be masked
 */
export default function mkCharNum(): MaskoseCharNum {
  return {
    type: MASKOSE_CHAR_NUM_TYPE,
    repetitions: 1,
    maskedValueLengthConditions: [],
    valueToBeMaskedLengthConditions: [],
    regExp: /[0-9]/
  };
}