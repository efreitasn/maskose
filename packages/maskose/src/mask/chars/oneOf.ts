import { MaskoseChar, MaskoseCharBase } from '.';

export const MASKOSE_CHAR_ONE_OF_TYPE = 'MASKOSE_CHAR_ONE_OF_TYPE';

export interface MaskoseCharOneOf extends MaskoseCharBase {
  type: typeof MASKOSE_CHAR_ONE_OF_TYPE;
  chars: MaskoseChar[];
};

export default function mkCharOneOf(
  chars: MaskoseChar[]
): MaskoseCharOneOf {
  // TODO
  // Don't allow characters whose repetitions > 1 to be put here

  return {
    type: MASKOSE_CHAR_ONE_OF_TYPE,
    chars,
    repetitions: 1,
    valueLengthConditions: []
  };
}