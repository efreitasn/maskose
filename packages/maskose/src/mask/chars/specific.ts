import { MaskoseMaskCharBase } from '.';
import escapeRegExpChar from '../../utils/escapeRegExpChar';

export const MASKOSE_CHAR_SPECIFIC_TYPE = 'MASKOSE_CHAR_SPECIFIC_TYPE';
export const MASKOSE_CHAR_SPECIFIC_STR_ID = 'SPECIFIC';

export interface MaskoseMaskCharSpecific extends MaskoseMaskCharBase {
  type: typeof MASKOSE_CHAR_SPECIFIC_TYPE;
  char: string;
  regExp: RegExp;
};

/**
 * A specific character expected to be in the value to be masked
 */
export default function mkCharSpecific(char: string): MaskoseMaskCharSpecific {
  if (char.length === 0 || char.length > 1) {
    throw new Error('The provided character must have a size equal to 1');
  }

  const regExp = new RegExp(
    escapeRegExpChar(char)
  );

  return {
    type: MASKOSE_CHAR_SPECIFIC_TYPE,
    char,
    repetitions: 1,
    maskedValueLengthConditions: [],
    valueToBeMaskedLengthConditions: [],
    regExp
  };
}