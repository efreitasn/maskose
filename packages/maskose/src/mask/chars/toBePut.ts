import { MaskoseCharBase } from '.';

export const MASKOSE_CHAR_TO_BE_PUT_TYPE = 'MASKOSE_CHAR_TO_BE_PUT_TYPE';
export const MASKOSE_CHAR_TO_BE_PUT_STR_ID = 'TO_BE_PUT';

export interface MaskoseCharToBePut extends MaskoseCharBase {
  type: typeof MASKOSE_CHAR_TO_BE_PUT_TYPE;
  char: string;
};

/**
 * A character that's expected to be put in the mask when formatting instead of being a character
 * expected to be in the value to be masked. For example, the dash (-) in a phone number mask is
 * such a character
 * @param char Character to be put in the mask
 */
export default function mkCharToBePut(char: string): MaskoseCharToBePut {
  if (char.length === 0 || char.length > 1) {
    throw new Error('The provided character must have a size equal to 1');
  }

  return {
    type: MASKOSE_CHAR_TO_BE_PUT_TYPE,
    char,
    repetitions: 1,
    maskedValueLengthConditions: [],
    valueToBeMaskedLengthConditions: []
  };
}