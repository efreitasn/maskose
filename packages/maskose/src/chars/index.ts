import { MaskoseCharLetter } from './primitives/letter';
import { MaskoseCharSpecific } from './primitives/specific';
import { MaskoseCharNum } from './primitives/num';
import { MaskoseCharToBePut } from './primitives/toBePut';
import { MaskoseCharRepeat } from './nonPrimitives/repeat';
import { MaskoseCharPredicateFn } from './nonPrimitives/predicateFn';

export { default as mkCharLetter } from './primitives/letter';
export { default as mkCharNum } from './primitives/num';
export { default as mkCharSpecific } from './primitives/specific';
export { default as mkCharToBePut } from './primitives/toBePut';
export { default as mkCharRepeat } from './nonPrimitives/repeat';
export { default as mkCharPredicateFn } from './nonPrimitives/predicateFn';

export type MaskoseChar =
  | MaskoseCharLetter
  | MaskoseCharSpecific
  | MaskoseCharNum
  | MaskoseCharToBePut
  | MaskoseCharRepeat
  | MaskoseCharPredicateFn
;

export type PrimitiveMaskoseChar =
  | MaskoseCharLetter
  | MaskoseCharSpecific
  | MaskoseCharNum
  | MaskoseCharToBePut
;

export type NonPrimitiveMaskoseChar = Exclude<MaskoseChar, PrimitiveMaskoseChar>;