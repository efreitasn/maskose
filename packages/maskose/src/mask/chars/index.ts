import { MaskoseCharLetter } from './letter';
import { MaskoseCharSpecific } from './specific';
import { MaskoseCharNum } from './num';
import { MaskoseCharToBePut } from './toBePut';
import { MaskoseCharGroup } from './group';

export const MASKOSE_CHAR_VALUE_LENGTH_CONDITION_GREATER_THAN = 'gt';
export const MASKOSE_CHAR_VALUE_LENGTH_CONDITION_EQUAL_TO = 'eq';
export const MASKOSE_CHAR_VALUE_LENGTH_CONDITION_LESS_THAN = 'lt';

export type MaskoseCharValueLengthConditionType =
  | typeof MASKOSE_CHAR_VALUE_LENGTH_CONDITION_GREATER_THAN
  | typeof MASKOSE_CHAR_VALUE_LENGTH_CONDITION_EQUAL_TO
  | typeof MASKOSE_CHAR_VALUE_LENGTH_CONDITION_LESS_THAN
;

export type MaskoseCharValueLengthCondition = {
  type: MaskoseCharValueLengthConditionType;
  num: number;
};

export interface MaskoseCharBase {
  repetitions: number;
  valueToBeMaskedLengthConditions: MaskoseCharValueLengthCondition[];
  maskedValueLengthConditions: MaskoseCharValueLengthCondition[];
};

export type MaskoseChar =
  | MaskoseCharLetter
  | MaskoseCharSpecific
  | MaskoseCharNum
  | MaskoseCharToBePut
  | MaskoseCharGroup
;