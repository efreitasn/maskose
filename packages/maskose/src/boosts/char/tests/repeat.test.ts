import test from 'ava';
import mkCharBoostRepeat from '../repeat';
import { mkCharLetter } from '../../../chars';

test('should match the object returned by mkCharBoostRepeat()', t => {
  const mkCharBoostRepeatWithNum = mkCharBoostRepeat(4);
  const result = mkCharBoostRepeatWithNum(mkCharLetter());

  t.snapshot(result);
});