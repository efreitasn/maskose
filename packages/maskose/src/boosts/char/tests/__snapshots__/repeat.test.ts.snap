// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the object returned by mkCharBoostRepeat() 1`] = `
Object {
  "maskedValueLengthConditions": Array [],
  "regExp": /\\[0-9\\]/,
  "repetitions": 5,
  "type": "MASKOSE_CHAR_NUM_TYPE",
  "valueToBeMaskedLengthConditions": Array [],
}
`;
